### ISSUE TITLE
Add `AttachScreenshot` option

### ISSUE THREAD

**Issue by bruno-garcia** [getsentry/sentry-unity#284]
When the option is set to true, events will include a screenshot of the game.

This feature is being added to Xamarin:

https://github.com/getsentry/sentry-xamarin/search?q=screenshot

https://github.com/getsentry/sentry-xamarin/blob/411be60d616789ad81ec13e2e50c7c2a777b4c67/Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotEventProcessor.droid.ios.cs

**Comment by bitsandfoxes** [comment]
Gotchas to keep in mind:
- Record at end of frame i.e. a coroutine yield return new WaitForEndOfFrame() and not in the middle of rendering
- ScreenCapture.CaptureScreenshotAsTexture() returns a Texture2D that can be encoded to .png, .jpg, .tga byte[]

**Comment by bruno-garcia** [comment]
Strategy on Xamarin changed:

https://github.com/getsentry/sentry-xamarin/blob/main/Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotAttachment.cs
https://github.com/getsentry/sentry-xamarin/blob/main/Src/Sentry.Xamarin/Internals/Device/Screenshot/ScreenshotAttachmentContent.cs

Would be nice to have a Sample so users can see the usage when they import the package.

Unmerged spec PR: https://github.com/getsentry/develop/pull/422

**Comment by oddgames-david** [comment]
> Gotchas to keep in mind:
> 
> * Record at end of frame i.e. a coroutine yield return new WaitForEndOfFrame() and not in the middle of rendering
> * ScreenCapture.CaptureScreenshotAsTexture() returns a Texture2D that can be encoded to .png, .jpg, .tga byte[]

Also:

- Native screenshot would be better, ensures always receiving a screenshot when the exception is threaded or from a java/objective-c error (so long as it didn't crash)
- Need to resize it before compressing to jpg, make the max height and compress amount a setting

This is my current implementation but most of the time I'd love to have screenshots for native exceptions, they are harder to diagnose.
```
    public class ScreenshotAttachment : IAttachmentContent
    {

        public Stream GetStream()
        {

            if (!UnityThread.IsRunningOnUnityThread)
                return new MemoryStream();

            float ratio = (float)Screen.width / (float)Screen.height;

            int targetHeight = Mathf.Min(Screen.height, 360);
            int targetWidth = Mathf.RoundToInt((float)targetHeight * ratio);

            RenderTexture screenshot = RenderTexture.GetTemporary(Screen.width, Screen.height);

            ScreenCapture.CaptureScreenshotIntoRenderTexture(screenshot);
            
            // Create a render texture to render into
            RenderTexture screenshotResize = RenderTexture.GetTemporary(targetWidth, targetHeight, 0);

            Graphics.Blit(screenshot, screenshotResize);

            RenderTexture.ReleaseTemporary(screenshot);

            RenderTexture previousActiveRT = RenderTexture.active;
            RenderTexture.active = screenshotResize;

            // Create a texture & read data from the active RenderTexture
            Texture2D result = new Texture2D(targetWidth, targetHeight) { name = "Screenshot" };
            result.ReadPixels(new Rect(0, 0, targetWidth, targetHeight), 0, 0);
            result.Apply();

            // Reset to initial state
            RenderTexture.active = previousActiveRT;

            RenderTexture.ReleaseTemporary(screenshotResize);

            var stream = new MemoryStream(result.EncodeToJPG(80));

            result.SafeDestroy();

            return stream;
        }

    }
```

**Comment by bitsandfoxes** [comment]
The resizing of the screenshot is a good catch. Thanks for sharing!
The native screenshot is a good idea and something that's planned to be added.

**Comment by oddgames-david** [comment]
If you allowed attachment streams to be async we could implement screenshots fairly easily ourselves. Or even just give us a callback 1 frame before you actually call GetStream

**Comment by pistoleta** [comment]
> > Gotchas to keep in mind:
> > 
> > * Record at end of frame i.e. a coroutine yield return new WaitForEndOfFrame() and not in the middle of rendering
> > * ScreenCapture.CaptureScreenshotAsTexture() returns a Texture2D that can be encoded to .png, .jpg, .tga byte[]
> 
> Also:
> 
> * Native screenshot would be better, ensures always receiving a screenshot when the exception is threaded or from a java/objective-c error (so long as it didn't crash)
> * Need to resize it before compressing to jpg, make the max height and compress amount a setting
> 
> This is my current implementation but most of the time I'd love to have screenshots for native exceptions, they are harder to diagnose.
> 
> ```
>     public class ScreenshotAttachment : IAttachmentContent
>     {
> 
>         public Stream GetStream()
>         {
> 
>             if (!UnityThread.IsRunningOnUnityThread)
>                 return new MemoryStream();
> 
>             float ratio = (float)Screen.width / (float)Screen.height;
> 
>             int targetHeight = Mathf.Min(Screen.height, 360);
>             int targetWidth = Mathf.RoundToInt((float)targetHeight * ratio);
> 
>             RenderTexture screenshot = RenderTexture.GetTemporary(Screen.width, Screen.height);
> 
>             ScreenCapture.CaptureScreenshotIntoRenderTexture(screenshot);
>             
>             // Create a render texture to render into
>             RenderTexture screenshotResize = RenderTexture.GetTemporary(targetWidth, targetHeight, 0);
> 
>             Graphics.Blit(screenshot, screenshotResize);
> 
>             RenderTexture.ReleaseTemporary(screenshot);
> 
>             RenderTexture previousActiveRT = RenderTexture.active;
>             RenderTexture.active = screenshotResize;
> 
>             // Create a texture & read data from the active RenderTexture
>             Texture2D result = new Texture2D(targetWidth, targetHeight) { name = "Screenshot" };
>             result.ReadPixels(new Rect(0, 0, targetWidth, targetHeight), 0, 0);
>             result.Apply();
> 
>             // Reset to initial state
>             RenderTexture.active = previousActiveRT;
> 
>             RenderTexture.ReleaseTemporary(screenshotResize);
> 
>             var stream = new MemoryStream(result.EncodeToJPG(80));
> 
>             result.SafeDestroy();
> 
>             return stream;
>         }
> 
>     }
> ```

How do you attach this to  the event?

**Comment by pistoleta** [comment]
> If you allowed attachment streams to be async we could implement screenshots fairly easily ourselves. Or even just give us a callback 1 frame before you actually call GetStream

If BeforeSend would be async we could also set attachments in the context without problems no?

**Comment by pistoleta** [comment]
To be honest I never expected Sentry would not have a feature like this... Unity Diagnostics doesn't have 1/10 part of the features you own but does have the screenshot... I dont mean to criticize is just a screenshot of the game just when the exception happened is extremely useful and might save you a lot of time finding where a problem is.

I hope you guys consider giving some priority to this or a temporary workaround.
Thanks a lot!

**Comment by bruno-garcia** [comment]
This is a priority to us. We're adding support for screenshots on iOS and Android, and there's a thumbnail in Sentry's issue detail page.
The spec for SDK development is here: https://develop.sentry.dev/sdk/features/#screenshots and we plan on adding this to Unity too

**Comment by oddgames-david** [comment]
@pistoleta Here's an updated version that works some of the time. You just add it to your Configure(SentryUnityOptions options) method.

`
public class ScreenshotAttachment : IAttachmentContent, UnityEngine.ILogHandler
{
    private ILogHandler handler;
    private Texture2D screenshot;

    public ScreenshotAttachment()
    {
        handler = Debug.unityLogger.logHandler;
        Debug.unityLogger.logHandler = this;
    }

    private async Task Capture()
    {
        
        Debug.Log("Capturing screenshot");

        await new WaitForEndOfFrame();

        float ratio = (float)Screen.width / (float)Screen.height;

        int targetHeight = Mathf.Min(Screen.height, 360);
        int targetWidth = Mathf.RoundToInt((float)targetHeight * ratio);

        RenderTexture screenshotRt = RenderTexture.GetTemporary(Screen.width, Screen.height);

        ScreenCapture.CaptureScreenshotIntoRenderTexture(screenshotRt);

        // Create a render texture to render into
        RenderTexture screenshotResize = RenderTexture.GetTemporary(targetWidth, targetHeight, 0);

#if UNITY_IOS || UNITY_EDITOR
        Graphics.Blit(screenshotRt, screenshotResize, new Vector2(1, -1), new Vector2(0, 1));
#else
        Graphics.Blit(screenshotRt, screenshotResize);
#endif
        RenderTexture.ReleaseTemporary(screenshotRt);

        RenderTexture previousActiveRT = RenderTexture.active;
        RenderTexture.active = screenshotResize;

        // Create a texture & read data from the active RenderTexture
        screenshot = new Texture2D(targetWidth, targetHeight) { name = "Screenshot" };
        screenshot.ReadPixels(new Rect(0, 0, targetWidth, targetHeight), 0, 0);
        screenshot.Apply();
    }

    public Stream GetStream()
    {


        if (screenshot == null)
        {
            UnityEngine.Debug.Log("Returning empty stream due to missing screenshot texture2d");
            return new MemoryStream();
        }

        var stream = new MemoryStream(screenshot.EncodeToJPG(80));

        screenshot.SafeDestroy();
        UnityEngine.Debug.Log("captured screenshot attachment");
        return stream;
    }

    public void LogFormat(LogType logType, UnityEngine.Object context, string format, params object[] args)
    {
        handler.LogFormat(logType, context, format, args);
    }

    public async void LogException(Exception exception, UnityEngine.Object context)
    {
        await Capture();
        handler.LogException(exception,context);
    }
}`

**Comment by vaind** [comment]
Hmm, seeing the screenshots don't have cursor (when in UI), maybe we should also include cursor position in all events? https://docs.unity3d.com/ScriptReference/Input-mousePosition.html 

It's not the same issue though, just asking here and if it makes sense we can have a followup issue for that.

**Comment by vaind** [comment]
For backup: an async version would look like this

```
        ScreenshotAttachmentContent.GetStream():
        {
            if (!_behaviour.MainThreadData.IsMainThread())
            {
                _options.DiagnosticLogger?.LogDebug("Won't capture screenshot because we're not on the main thread");
                return new MemoryStream();
            }
            else
            {
                return new ScreenshotCaptureStream(_options, _behaviour);
            }
        }

    // The async version, currently unused because we want the screenshot to be as early as possible after the error
    internal class ScreenshotCaptureStream : Stream
    {
        private readonly SentryOptions _options;
        private readonly TaskCompletionSource<byte[]> _tcs;

        public ScreenshotCaptureStream(SentryOptions options, SentryMonoBehaviour behaviour)
        {
            _options = options;
            _tcs = new();
            behaviour.StartCoroutine(CaptureScreenshot());
        }

        private IEnumerator CaptureScreenshot()
        {
            _options.DiagnosticLogger?.LogDebug("Capturing screenshot");

            yield return new WaitForEndOfFrame();

            var texture = ScreenCapture.CaptureScreenshotAsTexture();
            byte[] bytes = texture.EncodeToJPG();
            UnityEngine.Object.Destroy(texture);

            _options.DiagnosticLogger?.LogDebug("Screenshot captured: {0} bytes", bytes.Length);

            _tcs.TrySetResult(bytes);
        }

        public override async Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
        {
            var bytes = await _tcs.Task.ConfigureAwait(false);
            _options.DiagnosticLogger?.LogDebug("Awaiting screenshot finished: {0} bytes", bytes.Length);
            var writer = new BinaryWriter(destination);
            {
                writer.Write(bytes);
            }
            _options.DiagnosticLogger?.LogDebug("Finished writing the screenshot");
        }

        public override bool CanRead => throw new NotImplementedException();
        public override bool CanSeek => throw new NotImplementedException();
        public override bool CanWrite => throw new NotImplementedException();
        public override long Length => throw new NotImplementedException();
        public override long Position { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
        public override void Flush() => throw new NotImplementedException();
        public override int Read(byte[] buffer, int offset, int count) => throw new NotImplementedException();
        public override long Seek(long offset, SeekOrigin origin) => throw new NotImplementedException();
        public override void SetLength(long value) => throw new NotImplementedException();
        public override void Write(byte[] buffer, int offset, int count) => throw new NotImplementedException();
    }
```

**Comment by pistoleta** [comment]
I understand for this you had to implement a IDiagnosticLogger and assign it to optios.DiagnosticLogger inside the SentryOptionsConfiguration, correct?

**Comment by vaind** [comment]
No that's not the case. You can see the actual implementation in #670 - ScreenShotAttachment.cs

**Comment by pistoleta** [comment]
> No that's not the case. You can see the actual implementation in #670 - ScreenShotAttachment.cs

I still fail to see how the example in GitHub you pointed can be added to the SentryOptionsConfiguration. I tested in editor and yes, it sends a screenshot but I thought we were talking about attaching a screenshot the moment an exception/error triggers... am I missing something?
I apologize if im being too dumb.

**Comment by vaind** [comment]
It's not released yet so it wouldn't work for you yet, unless you've made a custom build with that branch (which I strongly assume you haven't, based on your message).

I'd suggest waiting for the next package release (v0.14.0) which should include this feature, if everything goes according to the plan (i.e. we don't find any blocking issues)

**Comment by pistoleta** [comment]
> It's not released yet so it wouldn't work for you yet, unless you've made a custom build with that branch (which I strongly assume you haven't, based on your message).
> 
> I'd suggest waiting for the next package release (v0.14.0) which should include this feature, if everything goes according to the plan (i.e. we don't find any blocking issues)

Ah ok, thanks for your kindness and understanding, GitHub it's a bit hard to follow but I'll get used to it.

**Comment by oddgames-david** [comment]
@vaind I'd watch out for black screenshots when not using

`yield new WaitForEndOfFrame();`

**Comment by bitsandfoxes** [comment]
> I downloaded last release and used the example for sending screenshot but the screenshot seems to be captured upwards.... any hint where the problem might be? I used the exact same code provided in the example.

Which platform did you target?

**Comment by pistoleta** [comment]
iOS, btw i had to comment:
`result.SafeDestroy();` 
since Texture2D doesn't have that method implemented. What's this line supposed to do, release it from memory?

**Comment by vaind** [comment]
> used the example for sending screenshot

Do you mean the example code, not the new functionality implemented by the latest version? That example is old and doesn't do proper mirroring (necessary for most platforms)

**Comment by bitsandfoxes** [comment]
Sorry to see you run into problems.
Could you open up a new issue detailing what happened, where (platform, Unity-version), and what you had to do to get there/fix it? The `Attach Screenshot` feature is supposed to work out of the box and you definitely should not have to comment anything to make it work.

**Comment by bitsandfoxes** [comment]
If you're updating the Sentry package you also have to update the samples coming with it.

**Comment by vaind** [comment]
Also, there's no `result.SafeDestroy();` anywhere in this repository as far as I can tell - what example are you actually referring to?

**Comment by pistoleta** [comment]
wrong script 🤦  , seems I already had a ScreenShotAttachement that tried to make work some time ago. Apologies

**Comment by bitsandfoxes** [comment]
No worries! Happy to see it worked out for you regardless!

**Comment by pistoleta** [comment]
sorry but I dont see any way to attach screenshot in the 0.15.0 at least nothing new in the examples

**Comment by vaind** [comment]
you can enable it in the settings in the Unity UI (Tools->Sentry)

**Comment by pistoleta** [comment]
awesome, sorry I didnt see it in the changelog

**Comment by pistoleta** [comment]
sorry but I dont see any image attachment, neither from iOS reports or if I send one from unity editor... I enabled from the settings In tools->sentry and also I the SentryOptionsCOnfiguration ScriptableObject.
